### PR TITLE
Fixed bug in response mapping

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-260
Slack: https://keboolaglobal.slack.com/archives/C0553EM2RGX/p1691058230258039

Changes:
- Fixed bug in response mapping: 
  - `type ColumnMetadata map[string]MetadataDetail`
  - :arrow_right: 
  - `type ColumnMetadata map[string][]MetadataDetail`
- Improved tests.
  - Added operation: `CreateOrUpdateTableMetadata`
    - It will be used also in Stream UI backend changes, to mark tables used by the Stream API (for UI).

### CODE REVIEW

File: pkg/keboola/storage_table.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ type SourceProject struct {
 	Name string    `json:"name"`
 }
 
-type ColumnMetadata map[string]MetadataDetail
+type ColumnMetadata map[string][]MetadataDetail
```
Comment (michaljurecko) [inline pkg/keboola/storage_table.go]: Fixnuty bug, zvysok commitu su vylepsenia v namingu a komentaroch (tato cast bola nejaka zanedbana).

File: pkg/keboola/storage_table.go [hunk 2]
```diff
@@ -259,11 +259,11 @@ func (a *API) CreateTableFromFileRequest(tableID TableID, dataFileID int, opts .
 			return a.WaitForStorageJob(waitCtx, job)
 		}).
 		WithOnSuccess(func(_ context.Context, _ request.HTTPResponse) error {
-			bytes, err := jsonLib.Marshal(job.Results)
+			resultBytes, err := jsonLib.Marshal(job.Results)
```
Comment (michaljurecko) [inline pkg/keboola/storage_table.go:248]: `bytes` je nazov pouziteho pkg, IDE na to upozornuje.

File: pkg/keboola/storage_table.go [hunk 3]
```diff
@@ -402,9 +402,9 @@ type TableUnloadRequestBuilder struct {
 type UnloadFormat string
 
 const (
-	// CSV formatted according to RFC4180. This is the default format.
+	// UnloadFormatCSV generates CSV formatted according to RFC4180. This is the default format.
```
Comment (michaljurecko) [inline pkg/keboola/storage_table.go:391]: Komentare tohto typu maju zacinat nazvom objektu, ku ktoremu sa viazu (IDE na to upozornuje).

File: pkg/keboola/old_queue_test.go [hunk 4]
```diff
@@ -14,7 +14,7 @@ func TestCreateOldQueueJobRequest(t *testing.T) {
 	t.Parallel()
 
 	ctx := context.Background()
-	api := APIClientForAnEmptyProject(t, ctx, testproject.WithQueueV1())
+	_, api := APIClientForAnEmptyProject(t, ctx, testproject.WithQueueV1())
```
Comment (michaljurecko) [inline pkg/keboola/old_queue_test.go:17]: Upravena funkcia `APIClientForAnEmptyProject`, vrati uz aj `project`, potrebujem `project.StorageAPIHost()`.

`project` obsahuje informacie o pouzitom testovacom projekte.

File: pkg/keboola/storage_table.go [hunk 5]
```diff
@@ -47,20 +47,6 @@ type SourceProject struct {
 	Name string    `json:"name"`
 }
 
-type ColumnMetadata map[string][]MetadataDetail
```
Comment (michaljurecko) [inline pkg/keboola/storage_table.go]: Presunute do `storage_table_metadata.go`.

File: pkg/keboola/storage_table_metadata.go [hunk 6]
```diff
@@ -1 +1,65 @@
...
+		*r = ColumnsMetadata{}
+		return nil
+	}
+	// see https://stackoverflow.com/questions/43176625/call-json-unmarshal-inside-unmarshaljson-function-without-causing-stack-overflow
+	type _r ColumnsMetadata
+	return jsonLib.Unmarshal(data, (*_r)(r))
+}
+
+// CreateOrUpdateTableMetadata https://keboola.docs.apiary.io/#reference/metadata/table-metadata/create-or-update
+func (a *API) CreateOrUpdateTableMetadata(tableID TableID, provider string, tableMetadata []TableMetadataRequest, columnsMetadata []ColumnMetadataRequest) request.APIRequest[*TableMetadataResponse] {
```
Comment (michaljurecko) [inline pkg/keboola/storage_table_metadata.go:45]: Pridany API call, aby som mohol bug fix otestovat.

File: pkg/keboola/storage_table_test.go [hunk 7]
```diff
@@ -290,6 +290,111 @@ func TestTableApiCalls(t *testing.T) {
 	}
 	assert.True(t, tableFound)
 
+	// Get table (without table and columns metadata)
```
Comment (michaljurecko) [inline pkg/keboola/storage_table_test.go:293]: Get table, bez metadat.

File: pkg/keboola/storage_table_test.go [hunk 8]
```diff
@@ -290,6 +290,111 @@ func TestTableApiCalls(t *testing.T) {
...
+			},
+			[]ColumnMetadataRequest{
+				{ColumnName: "first", Key: "columnMetadata1", Value: "value3"},
+				{ColumnName: "first", Key: "columnMetadata2", Value: "value4"},
+				{ColumnName: "second", Key: "columnMetadata3", Value: "value5"},
+				{ColumnName: "second", Key: "columnMetadata4", Value: "value6"},
+			},
+		).
+		Send(ctx)
+	assert.NoError(t, err)
```
Comment (michaljurecko) [inline pkg/keboola/storage_table_test.go:336]: Nastavenie metadat.

File: pkg/keboola/storage_table_test.go [hunk 9]
```diff
@@ -290,6 +290,111 @@ func TestTableApiCalls(t *testing.T) {
...
+				{Key: "columnMetadata2", Value: "value4", Provider: "go-client-test"},
+			},
+			"second": {
+				{Key: "columnMetadata3", Value: "value5", Provider: "go-client-test"},
+				{Key: "columnMetadata4", Value: "value6", Provider: "go-client-test"},
+			},
+		},
+	}, resMetadata)
+
+	// Get table (with table and columns metadata)
```
Comment (michaljurecko) [inline pkg/keboola/storage_table_test.go:358]: Get table s metadatami.

File: pkg/keboola/storage_table.go (hunk 1 above)
Comment (michaljurecko) [reply pkg/keboola/storage_table.go]: Ostatne commits riesia, ako to otestovat.

**Review by zajca — APPROVED** [review]
Vypadá to cajk, zajímavé že sa to neobjevilo v KaaC

**Comment by michaljurecko** [comment]
@zajca lebo pri zakladani novej tabulky (noveho exportu) tam ziadne metadata nemas, v tickete sa snazia pouzit existujucu + ma metadata.

**Pull request merged**